jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/PathMappingsHandler.java

```diff
@@ -111,11 +113,48 @@ public boolean handle(Request request, Response response, Callback callback) thr
             return false;
         }
         Handler handler = matchedResource.getResource();
+        PathSpec pathSpec = matchedResource.getPathSpec();
         if (LOG.isDebugEnabled())
             LOG.debug("Matched {} to {} -> {}", pathInContext, matchedResource.getPathSpec(), handler);
-        boolean handled = handler.handle(request, response, callback);
+
+        PathSpecRequest pathSpecRequest = new PathSpecRequest(request, pathSpec);
+        boolean handled = handler.handle(pathSpecRequest, response, callback);
         if (LOG.isDebugEnabled())
             LOG.debug("Handled {} {} by {}", handled, pathInContext, handler);
         return handled;
     }
+
+    private static class PathSpecRequest extends Request.Wrapper
+    {
+        private final PathSpec pathSpec;
+        private final Context context;
+
+        public PathSpecRequest(Request request, PathSpec pathSpec)
+        {
+            super(request);
+            this.pathSpec = pathSpec;
+            setAttribute(PathSpec.class.getName(), this.pathSpec);
+            this.context = new Context.Wrapper(request.getContext())
```

Comment:
It might be good to reuse this wrapper for every  mapping to the same context. ... but maybe don't do this yet... just keep it in mind whilst we work out the other details.

Reply:
Can't be done... capturing lambda